lib/graphql/analysis/ast/query_complexity.rb

```diff
@@ -65,109 +53,76 @@ def on_enter_field(node, parent, visitor)
           # we'll visit them when we hit the spreads instead
           return if visitor.visiting_fragment_definition?
           return if visitor.skipping?
-          complexities_on_type = @complexities_on_type_by_query[visitor.query]
-          complexities_on_type.pop
+          scopes_stack = @complexities_on_type_by_query[visitor.query]
+          scopes_stack.pop
         end
 
         private
 
         # @return [Integer]
         def max_possible_complexity
-          @complexities_on_type_by_query.reduce(0) do |total, (query, complexities_on_type)|
-            root_complexity = complexities_on_type.last
-            # Use this entry point to calculate the total complexity
-            total_complexity_for_query = merged_max_complexity_for_scopes(query, [root_complexity.scoped_children])
-            total + total_complexity_for_query
+          @complexities_on_type_by_query.reduce(0) do |total, (query, scopes_stack)|
+            total + merged_max_complexity_for_scopes(query, [scopes_stack.first])
           end
         end
 
         # @param query [GraphQL::Query] Used for `query.possible_types`
-        # @param scoped_children_hashes [Array<Hash>] Array of scoped children hashes
+        # @param scopes [Array<ScopedTypeComplexity>] Array of scoped type complexities
         # @return [Integer]
-        def merged_max_complexity_for_scopes(query, scoped_children_hashes)
-          # Figure out what scopes are possible here.
+        def merged_max_complexity_for_scopes(query, scopes)
+          # Aggregate a set of all possible scope types encountered (scope keys).
           # Use a hash, but ignore the values; it's just a fast way to work with the keys.
-          all_scopes = {}
-          scoped_children_hashes.each do |h|
-            all_scopes.merge!(h)
+          possible_scope_types = scopes.each_with_object({}) do |scope, memo|
+            memo.merge!(scope)
           end
 
-          # If an abstract scope is present, but _all_ of its concrete types
-          # are also in the list, remove it from the list of scopes to check,
-          # because every possible type is covered by a concrete type.
-          # (That is, there are no remainder types to check.)
-          prev_keys = all_scopes.keys
-          prev_keys.each do |scope|
-            next unless scope.kind.abstract?
-
-            missing_concrete_types = query.possible_types(scope).select { |t| !all_scopes.key?(t) }
-            # This concrete type is possible _only_ as a member of the abstract type.
-            # So, attribute to it the complexity which belongs to the abstract type.
-            missing_concrete_types.each do |concrete_scope|
-              all_scopes[concrete_scope] = all_scopes[scope]
+          # Expand abstract scope types into their concrete implementations;
+          # overlapping abstracts coalesce through their intersecting types.
+          possible_scope_types.keys.each do |possible_scope_type|
```

Comment:
For reducing allocations, could we use `each_key` here to avoid the intermediate array of keys? 

```suggestion
          possible_scope_types.each_key do |possible_scope_type|
```

Reply:
In this case, we need the intermediary array because the loop may modify `possible_scope_types`.

Reply:
Ah, right! Thanks.